yb-voyager/src/adaptiveparallelism/adaptive_parallelism.go

```diff
@@ -120,3 +133,50 @@ func getMaxCpuUsageInCluster(clusterMetrics map[string]tgtdb.NodeMetrics) (int,
 	}
 	return maxCpuPct, nil
 }
+
+/*
+Memory load is considered to be high in the following scenarios
+- Available memory of any node is less than 10% (MIN_AVAILABLE_MEMORY_THRESHOLD) of it's total memory
+- tserver root memory consumption of any node has breached it's soft limit.
+*/
+func isMemoryLoadHigh(clusterMetrics map[string]tgtdb.NodeMetrics) (bool, error) {
+	minMemoryAvailablePct := 100
+	isTserverRootMemorySoftLimitBreached := false
+	for _, nodeMetrics := range clusterMetrics {
+		if nodeMetrics.Status != "OK" {
+			continue
+		}
+		// check if tserver root memory soft limit is breached
+		tserverRootMemoryConsumption, err := strconv.ParseInt(nodeMetrics.Metrics["tserver_root_memory_consumption"], 10, 64)
+		if err != nil {
+			return false, fmt.Errorf("parsing tserver root memory consumption as int: %w", err)
+		}
+		tserverRootMemorySoftLimit, err := strconv.ParseInt(nodeMetrics.Metrics["tserver_root_memory_soft_limit"], 10, 64)
+		if err != nil {
+			return false, fmt.Errorf("parsing tserver root memory soft limit as int: %w", err)
+		}
+		if tserverRootMemoryConsumption > tserverRootMemorySoftLimit {
+			isTserverRootMemorySoftLimitBreached = true
+			break
+		}
+
+		// check if memory available is low
+		memoryAvailable, err := strconv.ParseInt(nodeMetrics.Metrics["memory_available"], 10, 64)
+		if err != nil {
+			return false, fmt.Errorf("parsing memory available as int: %w", err)
+		}
+		memoryTotal, err := strconv.ParseInt(nodeMetrics.Metrics["memory_total"], 10, 64)
+		if err != nil {
+			return false, fmt.Errorf("parsing memory total as int: %w", err)
+		}
+		if memoryAvailable == 0 || memoryTotal == 0 {
+			// invalid values
+			// on macos memory available is not available
+			continue
+		}
+		memoryAvailablePct := int((memoryAvailable * 100) / memoryTotal)
+		minMemoryAvailablePct = min(minMemoryAvailablePct, memoryAvailablePct)
```

Comment:
can we have a check here also like `isTserverRootMemorySoftLimitBreached` to break the loop.